### ISSUE TITLE
verify driver for gas-gauge on EVT1

### ISSUE THREAD

**Issue by networkimprov** [networkimprov/arch-packages#31]
(was a.a.k.007)

**Comment by mranostay** [comment]
Can confirm power_supply_changed() is the only visible uevent from the driver. Need to check the effort for battery percentage calculations.

**Comment by mranostay** [comment]
@networkimprov @tmlind Reviewing the other battery drivers in drivers/power and none use uevents to report battery levels.  Are we sure that we want to rely on uevents + udev?

Seems an daemon service polling the battery capacity property would be a better solution?

**Comment by networkimprov** [comment]
Under what conditions does the driver call power_supply_changed()? I imagined catching one of those events...

Does the driver already support sysfs_notify?

**Comment by mranostay** [comment]
On Tue, Sep 13, 2016 at 4:13 PM, Liam <notifications@github.com> wrote:

> Under what conditions does the driver call power_supply_changed()? I
> imagined catching one of those events...
>
Aha it is suppose to be reporting back in precent via uevents as is.. will
test this evening.

**Comment by networkimprov** [comment]
OK, but maybe it should report more often under 10%?

**Comment by mranostay** [comment]
Ah it is a poll interval of 6 minutes, not anything with 3% discharge.
Do you want to decrease the poll_interval?  
.....
static unsigned int poll_interval = 360;                                        
module_param(poll_interval, uint, 0644);                                        
MODULE_PARM_DESC(poll_interval,                                                 
                 "battery poll interval in seconds - 0 disables polling");     
.....

**Comment by networkimprov** [comment]
Ya, let's reduce that under 10% to 3min, and maybe again to 1min.

**Comment by mranostay** [comment]
Ok will add this functionality today.

**Comment by mranostay** [comment]
@tmlind @networkimprov  Patchset to enable configurable poll_interval rates for battery level

https://github.com/networkimprov/linux/commit/524bf5b08794abb354432996d55d1e0d4258d576.patch
https://github.com/networkimprov/linux/commit/524bf5b08794abb354432996d55d1e0d4258d576.patch

**Comment by networkimprov** [comment]
Is OK/preferable to make a new branch (off anvl-v4.7) for patches to a diff component :-)

**Comment by networkimprov** [comment]
I don't think we need to emit uevent if the battery level hasn't changed since last poll.

Can you draft a sample udev rule for the gauge? I gather it can adjust the poll_interval depending on battery capacity?

**Comment by mranostay** [comment]
Ok I'll look into the udev rule... not sure you can have conditionals greater or less than for battery capacity rule

**Comment by networkimprov** [comment]
I'll add a couple rules with wildcards to adjust down the poll_interval, just need to know how to set it, e.g.

```
KERNEL=="bq..." ACTION=="change" ATTR{voltage}=="3.[65]*" ATTR{poll_interval}="180" RUN+="/bin/sh -c 'echo DING > /dev/console'"
```

**Comment by mranostay** [comment]
On Thu, Sep 15, 2016 at 2:35 PM, Liam <notifications@github.com> wrote:

> I'll add a couple rules with wildcards to adjust down the poll_interval,
> just need to know how to set it, e.g.
>
> KERNEL=="bq..." ACTION=="change" ATTR{voltage}=="3.[65]*" ATTR{poll_interval}="180" RUN+="/bin/sh -c 'echo DING > /dev/console'"
>
>
poll_interval isn't a uevent property.. does that need to be reported as
well?

**Comment by mranostay** [comment]
On Thu, Sep 15, 2016 at 2:40 PM, Matt Ranostay <mranostay@gmail.com> wrote:

>
>
> On Thu, Sep 15, 2016 at 2:35 PM, Liam <notifications@github.com> wrote:
>
>> I'll add a couple rules with wildcards to adjust down the poll_interval,
>> just need to know how to set it, e.g.
>>
>> KERNEL=="bq..." ACTION=="change" ATTR{voltage}=="3.[65]*" ATTR{poll_interval}="180" RUN+="/bin/sh -c 'echo DING > /dev/console'"
>>
>>
> poll_interval isn't a uevent property.. does that need to be reported as
> well?
>

Forget udev looks at the sysfs properties as well

**Comment by networkimprov** [comment]
Can you test this rule idea...

```
KERNEL=="bq..." ACTION=="change" ATTR{voltage}=="3.[65]*" ATTR{poll_interval}="180" RUN+="/bin/sh -c 'echo DING > /dev/console'"
```

**Comment by mranostay** [comment]
On Fri, Sep 16, 2016 at 12:18 PM, Liam <notifications@github.com> wrote:

> Can you test this rule idea...
>
Ok will do..  I think the voltage needs to be voltage_now and it isn't a
floating point but in microvolts

**Comment by networkimprov** [comment]
Should match uV: {voltage}=="3[654]?????"

**Comment by mranostay** [comment]
KERNEL=="bq27425-0" ACTION=="change" ATTRS{voltage_now}=="3[56]*" RUN+="/bin/sh -c 'echo DING > /dev/console'"

Works. However dropped the ATTRS{poll_interval} because that will not exist because  /sys/module/bq27xxx_battery/parameters/poll_interval  is the correct place since it already exists

@tmlind Please review. https://github.com/networkimprov/linux/commit/cba20bc11b8a3c6a469e2401a3670201f8d12e15.patch

**Comment by networkimprov** [comment]
We need to write poll_interval from the rule. Can we move the existing sysfs entity, or add a sysfs entity which is an alias for it?

**Comment by mranostay** [comment]
Can't do?

KERNEL=="bq27425-0" ACTION=="change" ATTRS{voltage_now}=="3[56]*" RUN+="/bin/sh -c 'echo 10 > /sys/module/bq27xxx_battery/parameters/poll_interval'"

**Comment by networkimprov** [comment]
Well, ya, but that's not what users expect; voltage_now is available; why not poll_interval?

**Comment by mranostay** [comment]
On Fri, Sep 16, 2016 at 6:41 PM, Liam <notifications@github.com> wrote:

> Well, ya, but that's not what users expect; voltage_now is available; why
> not poll_interval?
>

voltage_now is an actual device related to a changing quantity.

poll_interval is a kernel parameter isn't going to change without userspace
doing it, and I don't see anyone basing rules on what a ATTR{poll_interval}
is set too (and really shouldn't).  And much more importantly the sysfs way
will not be accepted upstream (at least I suspect so, maybe Tony has
thoughts on that) because a kernel parameter already exists.

**Comment by networkimprov** [comment]
I need it accessible in udev to assign to it, not read it; lots of rules do ATTR{thing}="x"

**Comment by mranostay** [comment]
On Fri, Sep 16, 2016 at 7:31 PM, Liam <notifications@github.com> wrote:

> I need it accessible in udev to assign to it, not read it; lots of rules
> do ATTR{thing}="x"
>
>
Ah of course... ok so the previous patchset is good enough for that then.
Having that use case maybe upstream will be okay... be nice to avoid any
technical debt though

**Comment by networkimprov** [comment]
Can you post patch you sent upstream to your bq27xxx branch (after reverting https://github.com/networkimprov/linux/commit/2aeea223) ?

The power-supply fix, if allowed upstream, can also go on this branch.

**Comment by mranostay** [comment]
Ok will do.

**Comment by networkimprov** [comment]
Re property UPDATE_INTERVAL when I apply these two
https://github.com/networkimprov/linux/commit/2626f43c (accepted upstream)
https://github.com/networkimprov/linux/compare/b2cc66eb...7bf8c2ef (new patches)

I get:
error: patch failed: drivers/power/bq27xxx_battery.c:1026
error: drivers/power/bq27xxx_battery.c: patch does not apply

**Comment by mranostay** [comment]
On Tue, Sep 20, 2016 at 8:16 PM, Liam <notifications@github.com> wrote:

> Re property UPDATE_INTERVAL when I apply these two
> networkimprov/linux@2626f43
> <https://github.com/networkimprov/linux/commit/2626f43c> (accepted
> upstream)
> networkimprov/linux@b2cc66e...7bf8c2e
> <https://github.com/networkimprov/linux/compare/b2cc66eb...7bf8c2ef> (new
> patches)
>
> I get:
> error: patch failed: drivers/power/bq27xxx_battery.c:1026
> error: drivers/power/bq27xxx_battery.c: patch does not apply
>
> Probably since I reverted that sysfs way in that branch as well.. we
probably want rebase interactive and drop the reverts and original patches
at some point.

**Comment by networkimprov** [comment]
Would you want to call the new poll_interval_param_set() in net set_property()?

**Comment by mranostay** [comment]
On Tue, Sep 20, 2016 at 8:33 PM, Liam <notifications@github.com> wrote:

> Would you want to call the new poll_interval_param_set() in net
> set_property()?
>

I don't, and this has nothing to do with that other method. What I am
saying is a simple cherry-pick to the anvl-4.7 branch isn't going to work.

Note the revert here of the sysfs way of updating
https://github.com/networkimprov/linux/commit/b2cc66eb60dbd4f78cffaf857e1ea7c4e1c61a10

**Comment by networkimprov** [comment]
You can fix the patch-apply issue when sending upstream; not important for me now.

Aren't poll_interval_param_set() and set_property() updating the same field?

**Comment by mranostay** [comment]
On Tue, Sep 20, 2016 at 8:49 PM, Liam <notifications@github.com> wrote:

> You can fix the patch-apply issue when sending upstream; not important for
> me now.
>
> Aren't poll_interval_param_set() and set_property() updating the same
> field?
>

Yes there are. Shouldn't need locking on the global... ideally it would be
in the data struct for the driver, but can't break the ABI.

Maybe Tony can comment on if this is the best way.

**Comment by networkimprov** [comment]
Ah, setting module param updates every instance of the driver, and new patch updates one instance. Why didncha say so! :-)

**Comment by networkimprov** [comment]
This file needs a description of the new param:
https://github.com/networkimprov/linux/blob/anvl-v4.7-bq27xxx/Documentation/power/power_supply_class.txt

There are two existing time parameters; maybe we should use "TIME_TO_UPDATE"?
```
TIME_TO_EMPTY - seconds left for battery to be considered empty (i.e. while battery powers a load)
TIME_TO_FULL - seconds left for battery to be considered full (i.e. while battery is charging)
```

**Comment by networkimprov** [comment]
Besides docs, looks good to go upstream :-)

If they complain that we are the only client for this option, I'm sure we can find other drivers that are the sole clients of different options!

**Comment by mranostay** [comment]
On Wed, Sep 21, 2016 at 11:50 AM, Liam <notifications@github.com> wrote:

> This file needs a description of the new param:
> https://github.com/networkimprov/linux/blob/anvl-
> v4.7-bq27xxx/Documentation/power/power_supply_class.txt
>
> There are two existing time parameters; maybe we should use
> "TIME_TO_UPDATE"?
>

Yeah only issue with using that is that assumes it is a countdown till the
next update...

**Comment by networkimprov** [comment]
True, and time_between_updates is a lil verbose. Perhaps stick with poll_interval since that's the module param?

**Comment by networkimprov** [comment]
fyi, I pushed patch already accepted upstream to anvl-v4.7

**Comment by networkimprov** [comment]
What do you think about reposting this patchset with a Documentation rev, and CC'ing Sebastian?
http://lkml.iu.edu/hypermail/linux/kernel/1609.2/04064.html

**Comment by mranostay** [comment]
Ok needs to be done. Better than poking him I guess :)

**Comment by networkimprov** [comment]
Yes :) Post docs commit on https://github.com/networkimprov/linux/commits/anvl-v4.7-bq27xxx

**Comment by networkimprov** [comment]
Proposed param docs:

```
TIME_TO_FULL - seconds left for battery ...
POLL_INTERVAL - seconds between I2C requests to retrieve battery
stats from chip.
```

https://github.com/networkimprov/linux/blob/anvl-v4.7-bq27xxx/Documentation/power/power_supply_class.txt

**Comment by mranostay** [comment]
On Mon, Sep 26, 2016 at 8:32 PM, Liam <notifications@github.com> wrote:

> Proposed param docs:
>
> TIME_TO_FULL - seconds left for battery ...
> POLL_INTERVAL - seconds between I2C requests to retrieve battery
> stats from chip.
>
>
We shouldn't say the interface.. it doesn't have to be I2C there are likely
SPI, and other interface fuel gauge parts.

**Comment by networkimprov** [comment]
OK

**Comment by networkimprov** [comment]
Unrelated to poll_interval... We used to apply this patch. The current code has just one compatible item and no #else section. Do we need this?

```
+#ifdef CONFIG_OF
+static const struct of_device_id bq27xxx_of_match[] = {
+       { .compatible = "ti,bq27200", },
+       { .compatible = "ti,bq27500", },
+       { .compatible = "ti,bq27425", },
+       { .compatible = "ti,bq27742", },
+       { .compatible = "ti,bq27518", },
+       { },
+};
+MODULE_DEVICE_TABLE(of, bq27xxx_of_match);
+#else
+static const struct of_device_id bq27xxx_of_match[] = {
+       { },
+};
+#endif
```

**Comment by mranostay** [comment]
On Tue, Sep 27, 2016 at 1:20 PM, Liam <notifications@github.com> wrote:

> Unrelated to poll_interval... We used to apply this patch. The current
> code has just one compatible item and no #else section. Do we need this?
>
Probably not.. but I'll check this afternoon

**Comment by mranostay** [comment]
On Tue, Sep 27, 2016 at 1:33 PM, Matt Ranostay <mranostay@gmail.com> wrote:

>
>
> On Tue, Sep 27, 2016 at 1:20 PM, Liam <notifications@github.com> wrote:
>
>> Unrelated to poll_interval... We used to apply this patch. The current
>> code has just one compatible item and no #else section. Do we need this?
>>
> Probably not.. but I'll check this afternoon
>

We could change the "ti,bq27425" in our device tree to "ti,bq27000" so we
don't have to worry about that patchset.

I suspect a) another version got upstreamed b) or the maintainer rathers to
keep the match table more generic unless a device needs the data pointer.

**Comment by networkimprov** [comment]
> Please review the same two patchsets from before now on the anvl-v4.7-bq27xxx branch. Only changes is the property name and two patchsets have been squashed.   Let me know if anything needs to be updated (maybe docs ?) before sending off v2

There's a grammar flaw in docs. How about:

POLL_INTERVAL - seconds between queries to retrieve stats from hardware.

**Comment by mranostay** [comment]
Sent from my iPhone

> On Sep 27, 2016, at 18:04, Liam <notifications@github.com> wrote:
> 
> Please review the same two patchsets from before now on the anvl-v4.7-bq27xxx branch. Only changes is the property name and two patchsets have been squashed. Let me know if anything needs to be updated (maybe docs ?) before sending off v2
> 
> There's a grammar flaw in docs. How about:
> 
> POLL_INTERVAL - seconds between queries to retrieve stats from hardware.
> 

Ok thought I recalled a comment about that issue. Will fix

**Comment by networkimprov** [comment]
My gauge claims to be charging when it's discharging. The first check, shortly after boot, is correct. Within an hour it was wrong. Hope this is a driver bug vs hw...

```
[root@anvl ~]# udevadm info /sys/class/power_supply/bq27425-0
P: /devices/platform/68000000.ocp/48072000.i2c/i2c-1/1-0055/power_supply/bq27425-0
E: DEVPATH=/devices/platform/68000000.ocp/48072000.i2c/i2c-1/1-0055/power_supply/bq27425-0
E: POWER_SUPPLY_CAPACITY=86
E: POWER_SUPPLY_CAPACITY_LEVEL=Normal
E: POWER_SUPPLY_CHARGE_FULL=1178000
E: POWER_SUPPLY_CHARGE_FULL_DESIGN=1340000
E: POWER_SUPPLY_CHARGE_NOW=1037000
E: POWER_SUPPLY_CURRENT_NOW=-47000
E: POWER_SUPPLY_MANUFACTURER=Texas Instruments
E: POWER_SUPPLY_NAME=bq27425-0
E: POWER_SUPPLY_PRESENT=1
E: POWER_SUPPLY_STATUS=Discharging <========== OK ==========<
E: POWER_SUPPLY_TECHNOLOGY=Li-ion
E: POWER_SUPPLY_TEMP=304
E: POWER_SUPPLY_VOLTAGE_NOW=3979000
E: SUBSYSTEM=power_supply

P: /devices/platform/68000000.ocp/48072000.i2c/i2c-1/1-0055/power_supply/bq27425-0
E: DEVPATH=/devices/platform/68000000.ocp/48072000.i2c/i2c-1/1-0055/power_supply/bq27425-0
E: POWER_SUPPLY_CAPACITY=82
E: POWER_SUPPLY_CAPACITY_LEVEL=Normal
E: POWER_SUPPLY_CHARGE_FULL=1178000
E: POWER_SUPPLY_CHARGE_FULL_DESIGN=1340000
E: POWER_SUPPLY_CHARGE_NOW=996000
E: POWER_SUPPLY_CURRENT_NOW=-54000
E: POWER_SUPPLY_MANUFACTURER=Texas Instruments
E: POWER_SUPPLY_NAME=bq27425-0
E: POWER_SUPPLY_PRESENT=1
E: POWER_SUPPLY_STATUS=Charging <========== WRONG ==========<
E: POWER_SUPPLY_TECHNOLOGY=Li-ion
E: POWER_SUPPLY_TEMP=264
E: POWER_SUPPLY_VOLTAGE_NOW=3975000
E: SUBSYSTEM=power_supply
```

**Comment by mranostay** [comment]
On Tue, Sep 27, 2016 at 11:45 PM, Liam <notifications@github.com> wrote:

> My gauge claims to be charging when it's discharging. The first check,
> shortly after boot, is correct. Within an hour it was wrong. Hope this is a
> driver bug vs hw...
>
Scary but I don't think any kernel changes we have done could cause this.

We need to check how the POWER_SUPPLY_STATUS state is set.... call be naive
but battery charging circuits (not the fuel gauge in this case) are rather
well validated because of lawsuits from fires.

Fuel gauge reporting charging doesn't exactly mean it is esp if is using a
metric like voltage change.. but late will check in the morning :)

**Comment by networkimprov** [comment]
Also, maybe poll_interval_param_set() should do nothing if new_val==current_val?

**Comment by mranostay** [comment]
On Wed, Sep 28, 2016 at 12:20 AM, Liam <notifications@github.com> wrote:

> Also, maybe poll_interval_param_set() should do nothing if
> new_val==current_val?
>

Makes sense we avoid an unneeded cancel on the workqueue and rerun

**Comment by networkimprov** [comment]
Re attributes, power_supply_prop_current_now is negative. Might be related to status Charging although the instance of Discharging also had negative current_now...

**Comment by mranostay** [comment]
@networkimprov  You had USB cable connected I assume?

**Comment by networkimprov** [comment]
USB is disconnected. The driver does curr = -curr for some devices, maybe it's misunderstanding our device type.

**Comment by networkimprov** [comment]
poll_interval_param_set() needs a separate patch to ignore new==current interval

**Comment by mranostay** [comment]
Looking at bq27xxx_battery_status() it seems we do need to keep "ti,bq27425" so as long we don't change that it should be working. But I'll check in that part of the code though

**Comment by networkimprov** [comment]
Maybe we should include all the of_device_id entries from that patch I mentioned?
In our case .chip fields should be assigned BQ27421.

(I assume you meant BQ27XXX_FLAG_DSC set in flags.)

**Comment by networkimprov** [comment]
Do we need a table of strings with the same order as this enum?
https://github.com/networkimprov/linux/blob/anvl-v4.7/include/linux/power/bq27xxx_battery.h#L4

**Comment by networkimprov** [comment]
The table in patch above appears to match enum added in this commit:
https://git.kernel.org/cgit/linux/kernel/git/stable/linux-stable.git/commit/include/linux/power/bq27xxx_battery.h?id=424cfde49acaf1426e90837961e8aead0238b11b

**Comment by mranostay** [comment]
On Wed, Sep 28, 2016 at 5:06 PM, Liam <notifications@github.com> wrote:

> The table in patch above appears to match enum added in this commit:
> https://git.kernel.org/cgit/linux/kernel/git/stable/linux-
> stable.git/commit/include/linux/power/bq27xxx_battery.h?id=
> 424cfde49acaf1426e90837961e8aead0238b11b
>
What we have in our tree matches what currently exists in the enum
definition upstream.

**Comment by networkimprov** [comment]
The upstream enum has 7 entries, and 1 of_match_table entry (added recently by someone with no prior commits to this driver). Tony's of_match_table (i.e. above patch) appears to match the enum at the time he wrote it. We clearly have the wrong ID in pdata->chip, and no ref to our chip in the table. I suspect all the chips (incl comments) mentioned in the enum should be listed in match table, prob with .data to assign pdata->chip the correct ID.

```
enum bq27xxx_chip {
	BQ27000 = 1, /* bq27000, bq27200 */
	BQ27010, /* bq27010, bq27210 */
	BQ27500, /* bq27500, bq27510, bq27520 */
	BQ27530, /* bq27530, bq27531 */
	BQ27541, /* bq27541, bq27542, bq27546, bq27742 */
	BQ27545, /* bq27545 */
	BQ27421, /* bq27421, bq27425, bq27441, bq27621 */
```

**Comment by mranostay** [comment]
On Wed, Sep 28, 2016 at 7:59 PM, Liam <notifications@github.com> wrote:

> The upstream enum has 7 entries, and 1 of_match_table entry (added
> recently by someone with no prior commits to this driver). Tony's
> of_match_table (i.e. above patch) appears to match the enum at the time he
> wrote it. We clearly have the wrong ID in pdata->chip, and no ref to our
> chip in the table. I suspect all the chips (incl comments) mentioned in the
> enum should be listed in match table, prob with .data to assign pdata->chip
> the correct ID.
>
Ok makes more sense. Will check that theory out..

**Comment by mranostay** [comment]
On Wed, Sep 28, 2016 at 8:09 PM, Matt Ranostay <mranostay@gmail.com> wrote:

>
>
> On Wed, Sep 28, 2016 at 7:59 PM, Liam <notifications@github.com> wrote:
>
>> The upstream enum has 7 entries, and 1 of_match_table entry (added
>> recently by someone with no prior commits to this driver). Tony's
>> of_match_table (i.e. above patch) appears to match the enum at the time he
>> wrote it. We clearly have the wrong ID in pdata->chip, and no ref to our
>> chip in the table. I suspect all the chips (incl comments) mentioned in the
>> enum should be listed in match table, prob with .data to assign pdata->chip
>> the correct ID.
>>
> Although I do agree the device tree lookup table should have a .data field
but this isn't the issue with the driver.

chip->id is being set to BQ27421 (aka 7) because the i2c_device_id table
gets walked through before the device tree one.  e.g. "ti,bq27425" matches
the "bq27425".
So we are using the correct enum value, and registers. Although should
check if bq27421 and bq27425 are really one to one register compatible
parts.

**Comment by networkimprov** [comment]
Ah, I overlooked the _i2c source file.

>@networkimprov @tmlind  Probably doesn't really need review for something this simple to go upstream. But -> https://github.com/networkimprov/linux/commit/10bb9cb6859f5db982d396b161a6c12dd9346bcc

Those are char* type; is == a valid test?

**Comment by mranostay** [comment]
On Thu, Sep 29, 2016 at 12:07 AM, Liam <notifications@github.com> wrote:

> Ah, I overlooked the _i2c source file.
>
> @networkimprov <https://github.com/networkimprov> @tmlind
> <https://github.com/tmlind> Probably doesn't really need review for
> something this simple to go upstream. But -> networkimprov/linux@10bb9cb
> <https://github.com/networkimprov/linux/commit/10bb9cb6859f5db982d396b161a6c12dd9346bcc>
>
>
> Those are char* type; is == a valid test?
>

Ah great void * didn't toss a compiler warning...
You are correct partly it should be   *(unsigned int *) kp->arg

Will fix in v2

**Comment by networkimprov** [comment]
Maybe strncmp() on result of param_get_uint() ?

**Comment by networkimprov** [comment]
The manuals for 27425 & 421 have the same flags register layout: 06 (high?) & 07 (low?). If 06 is high, the read logic is wrong:
```
lower = pdata->read(dev, reg);   // 06, may be the high byte
upper = pdata->read(dev, reg+1); // 07
return (upper << 8) | lower;     // 07 is made high byte
```

**Comment by mranostay** [comment]
On Thu, Sep 29, 2016 at 12:56 AM, Liam <notifications@github.com> wrote:

> Maybe strncmp() on result of param_get_uint() ?
>
Yeah I don't think that is the best way since you would need to define a
char buffer to hold the data anyway.

**Comment by mranostay** [comment]
On Thu, Sep 29, 2016 at 6:39 AM, Liam <notifications@github.com> wrote:

> The manuals for 27425 & 421 have the same flags register layout: 06
> (high?) & 07 (low?). If 06 is high, the read logic is wrong:
>
>
Yeah the datasheet isn't exactly clear but the data is little endian, but
how they show it in that table is confusing. So that isn't the issue....
also none of the parts would work.

**Comment by networkimprov** [comment]
Dropped Markku a note; we'll hear back late this evening.

**Comment by mranostay** [comment]
https://github.com/networkimprov/linux/commit/a1f017e7098a1d30691a26c222e89346152600a9

Version 2.. actually comparing integer values this time.

**Comment by networkimprov** [comment]
Tested this one? :-) I'd name tmp "orig_val".

**Comment by mranostay** [comment]
@networkimprov Yeah also tested the last one... seems it didn't give compiler warnings or error out because it was comparing pointer addresses :).

maybe prev would be better than orig_val for upstreaming.

**Comment by networkimprov** [comment]
prev_val sounds good.

**Comment by mranostay** [comment]
@networkimprov Will revise and submit upstream.. Although I suspect we may not hear feedback for a few days since we about to enter  a merge window, and ton of conferences coming up.

**Comment by networkimprov** [comment]
Could prob wait a bit on sending upstream til we know whether there's other fixes for this driver.

**Comment by networkimprov** [comment]
It's too bad there isn't a macro to do *(unsigned int*)kp->arg.
Just for sanity sake, maybe add
`dev_dbg(dev, "previous poll_interval was %u\n", prev_val);`
